overlord/snapstate/snapmgr.go

```diff
@@ -551,6 +556,45 @@ func (m *SnapManager) ensureUbuntuCoreTransition() error {
 	return nil
 }
 
+// ensureSnapConfineApparmor ensures that we have a valid apparmor
+// profile for snap-confine when we are in re-exec mode
+func (m *SnapManager) ensureSnapConfineApparmor() error {
+	if !release.OnClassic {
+		return nil
+	}
+
+	root := filepath.Join(dirs.SnapMountDir, "/core/current/")
+	snapConfineInCore, err := filepath.EvalSymlinks(filepath.Join(root, "/usr/lib/snapd/snap-confine"))
+	if err != nil {
+		return err
+	}
+
+	apparmorProfilePath := filepath.Join(dirs.SystemApparmorDir, strings.Replace(snapConfineInCore[1:], "/", ".", -1))
+	if osutil.FileExists(apparmorProfilePath) {
+		return nil
+	}
+
+	// FIXME2: cleanup old entries here
```

Comment:
I think you would want to cleanup old profile where the corresponding core snap is no longer on the system. Leaving the old ones might help the revert scenario but the code would likely be simpler if instead only ever have the current core snap profile and remove all the others (that happens to mirror what we do with /var/lib/snapd/apparmor/profiles where we only have the profile for the current revision).